codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -133,7 +136,35 @@ public ChannelFuture writeData(final ChannelHandlerContext ctx, final int stream
 
         // Hand control of the frame to the flow controller.
         flowController().addFlowControlled(ctx, stream,
-                new FlowControlledData(ctx, stream, data, padding, endOfStream, promise));
+                new FlowControlledByteBuf(ctx, stream, data, padding, endOfStream, promise));
+        return promise;
+    }
+
+    @Override
+    public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, FileRegion data,
+            int padding, boolean endOfStream, ChannelPromise promise) {
+        final Http2Stream stream;
```

Comment:
Share this code with the other writeData by passing in a `ReferenceCounted`? You can check if the future is done/failed to know if you should call `addFlowControlled`.